File: client/src/main/scala/stasis/client/collection/BackupMetadataCollector.scala
package stasis.client.collection

import org.apache.pekko.stream.Materializer
import stasis.client.analysis.{Checksum, Metadata}
import stasis.client.compression.Compression
import stasis.client.model.{EntityMetadata, SourceEntity}

import java.nio.file.Path
import scala.concurrent.Future

trait BackupMetadataCollector {
  def collect(entity: Path, existingMetadata: Option[EntityMetadata]): Future[SourceEntity]
}

object BackupMetadataCollector {
  class Default(checksum: Checksum, compression: Compression)(implicit mat: Materializer) extends BackupMetadataCollector {
    override def collect(entity: Path, existingMetadata: Option[EntityMetadata]): Future[SourceEntity] =
      Metadata.collectSource(
        checksum = checksum,
        compression = compression,
        entity = entity,
        existingMetadata = existingMetadata
      )
  }

  object Default {
    def apply(checksum: Checksum, compression: Compression)(implicit mat: Materializer): Default =
      new Default(checksum, compression)
  }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/collection/BackupMetadataCollector.kt
package stasis.client_android.lib.collection

import stasis.client_android.lib.analysis.Checksum
import stasis.client_android.lib.analysis.Metadata
import stasis.client_android.lib.compression.Compression
import stasis.client_android.lib.model.EntityMetadata
import stasis.client_android.lib.model.SourceEntity
import java.nio.file.Path

interface BackupMetadataCollector {
    suspend fun collect(entity: Path, existingMetadata: EntityMetadata?): SourceEntity

    class Default(private val checksum: Checksum, private val compression: Compression) : BackupMetadataCollector {
        override suspend fun collect(entity: Path, existingMetadata: EntityMetadata?): SourceEntity =
            Metadata.collectSource(
                checksum = checksum,
                compression = compression,
                entity = entity, existingMetadata = existingMetadata
            )
    }
}
